scripts/common/rook-upgrade.sh

```diff
@@ -95,6 +100,96 @@ function rook_upgrade_prompt() {
     confirmN
 }
 
+# rook_upgrade_storage_check verifies that enough disk space exists for the rook upgrade to complete successfully.
+function rook_upgrade_storage_check() {
+    local current_version="$1"
+    local desired_version="$2"
+
+    local archive_size=
+    archive_size="$(rook_upgrade_required_archive_size "$current_version" "$desired_version")"
+
+    local container_directory=
+    if [ -n "$DOCKER_VERSION" ]; then
+        container_directory="/var/lib/docker"
+    else
+        container_directory="/var/lib/containerd"
+    fi
+
+    # if $container_directory and /var/lib/kurl are on the same filesystem, we need to check that there is space for all of the files
+    if [ "$(df -P "$container_directory" | awk 'END{print $1}')" = "$(df -P /var/lib/kurl | awk 'END{print $1}')" ]; then
+        # in total, we need space for 5.5x the archive size, AND there must be 15% free space on the filesystem afterwards
+        local total_required_size=
+        total_required_size=$((archive_size * 11 / 2)) # 5.5x archive size, rounded to an integer
+
+        local free_kb=
+        local free_mb=
+        free_kb="$(df -P /var/lib/kurl | awk 'END{print $4}')"
+        free_mb="$((free_kb / 1024))"
+
+        local total_kb=
+        local total_mb=
+        total_kb="$(df -P /var/lib/kurl | awk 'END{print $2}')"
+        total_mb="$((total_kb / 1024))"
+
+        local available_mb=
+        available_mb="$((free_mb - total_mb * 3 / 20))" # free space, excluding 15% of the total
+
+        if [ "$available_mb" -lt "$total_required_size" ]; then
+            logWarn "Not enough disk space to upgrade Rook."
+            logWarn "You need at least $total_required_size MB of free space on the filesystem containing /var/lib/kurl and $container_directory - and to have 15%% free space after that to avoid image pruning."
+            logWarn "Currently, only $available_mb MB of free space is available before reaching 85%% capacity."
+            logWarn "If you have already loaded images or started this Rook upgrade, it is possible that less space will be required. Would you like to continue anyways?"
+            if ! confirmN; then
+                return 1
```

Comment:
Why we are not using bail here? WDYT about use bail?

Reply:
because for every other bit in this flow we're returning and not bailing

Reply:
```
    if ! rook_upgrade_prompt "$current_version" "$desired_version" ; then
        bail "Not upgrading Rook"
    fi

    if ! rook_upgrade_storage_check "$current_version" "$desired_version" ; then
        bail "Not upgrading Rook"
    fi
```

Reply:
But in the `confirmN` we do not bail. Here:

```
if ! confirmN; then
                return 1
```

Should we not as we do in the other places?

Reply:
I can change it to just be `confirmN`, which is what we do in the `rook_upgrade_prompt` section

It's the same thing though, just spelled out

Reply:
`rook_upgrade_storage_check` returns 1, and so there is a bail in the snippet I pasted

Reply:
I see I think that is fine 👍 
Many ways to do the same :-)